pkg/inventory/inventory-client.go

```diff
@@ -6,473 +6,402 @@ package inventory
-		return nil, fmt.Errorf("found %d inventory objects with inventory id %s", l, inv.ID())
-	}
-	return clusterInv, nil
+type UpdateOptions struct{}
+
+type GetOptions struct{}
+
+type ListOptions struct{}
+
+type DeleteOptions struct{}
+
+var _ Client = &UnstructuredClient{}
+
+// UnstructuredInventory implements Inventory while also tracking the actual
+// KRM object from the cluster. This enables the client to update the
+// same object and utilize resourceVersion checks.
+type UnstructuredInventory struct {
+	BaseInventory
+	// ClusterObj represents the KRM which was last fetched from the cluster.
+	// used by the client implementation to performs updates on the object.
+	ClusterObj *unstructured.Unstructured
 }
 
-func (cic *ClusterClient) getClusterInventoryObjsByLabel(ctx context.Context, inv Info) (object.UnstructuredSet, error) {
-	localInv, err := cic.invToUnstructuredFunc(inv)
-	if err != nil {
-		return nil, err
+func (ui *UnstructuredInventory) Info() Info {
+	if ui.ClusterObj == nil {
+		return SingleObjectInfo{}
 	}
-	if localInv == nil {
-		return nil, fmt.Errorf("retrieving cluster inventory object with nil local inventory")
-	}
-	localObj := object.UnstructuredToObjMetadata(localInv)
-	mapping, err := cic.getMapping(localInv)
-	if err != nil {
-		return nil, err
-	}
-	groupResource := mapping.Resource.GroupResource().String()
-	namespace := localObj.Namespace
-	label, err := retrieveInventoryLabel(localInv)
-	if err != nil {
-		return nil, err
+	return NewSingleObjectInfo(ui.ID(), client.ObjectKeyFromObject(ui.ClusterObj),
+		maps.Clone(ui.ClusterObj.GetLabels()), maps.Clone(ui.ClusterObj.GetAnnotations()))
+}
+
+func (ui *UnstructuredInventory) Name() string {
+	if ui.ClusterObj == nil {
+		return ""
 	}
-	labelSelector := fmt.Sprintf("%s=%s", common.InventoryLabel, label)
-	klog.V(4).Infof("inventory object fetch by label (group: %q, namespace: %q, selector: %q)", groupResource, namespace, labelSelector)
+	return ui.ClusterObj.GetName()
+}
 
-	uList, err := cic.dc.Resource(mapping.Resource).Namespace(namespace).List(ctx, metav1.ListOptions{
-		LabelSelector: labelSelector,
-	})
-	if err != nil {
-		return nil, err
+func (ui *UnstructuredInventory) Namespace() string {
+	if ui.ClusterObj == nil {
+		return ""
 	}
-	var invList []*unstructured.Unstructured
-	for i := range uList.Items {
-		invList = append(invList, &uList.Items[i])
+	return ui.ClusterObj.GetNamespace()
+}
+
+func (ui *UnstructuredInventory) ID() ID {
+	if ui.ClusterObj == nil {
+		return ""
 	}
-	return invList, nil
+	// Empty string if not set.
+	return ID(ui.ClusterObj.GetLabels()[common.InventoryLabel])
+}
+
+func (ui *UnstructuredInventory) NewEmptyInventory() Inventory {
+	return ui
+}
+
+var _ Inventory = &UnstructuredInventory{}
+
+// BaseInventory is a boilerplate struct that contains the basic methods
+// to implement Inventory. Can be extended for different inventory implementations.
+type BaseInventory struct {
```

Comment:
Looks like now that Inventory has an Info method, BaseInventory no longer implements the whole Inventory interface. That makes me wonder if AbstractInventory or InventoryContents would be a better name.

Reply:
I like InventoryContents. It makes sense as a partial impl without the metadata.